### PR TITLE
Added Buffer.drop

### PR DESCRIPTION
**Pull request by toots**

### CODE REVIEW

File: stdlib/buffer.mli [hunk 1]
```diff
@@ -158,6 +158,11 @@ val truncate : t -> int -> unit
   Raise [Invalid_argument] if [len < 0] or [len > length b].
   @since 4.05.0 *)
 
+val drop : t -> int -> unit
+(** [drop b len] drops the first [len] bytes of [b]
+  Raise [Invalid_argument] if [len < 0] or [len > length b].
```
Comment (dra27) [inline stdlib/buffer.mli]:
```suggestion
  @raise Invalid_argument if [len < 0] or [len > length b].
```
(I realise the one above is also incorrect, but at least we'll slowly converge towards modern ocamldoc comments!)

File: stdlib/buffer.ml [hunk 2]
```diff
@@ -350,6 +350,15 @@ let truncate b len =
     else
       b.position <- len
 
+let drop b len =
+    if len < 0 || len > b.position then
+      invalid_arg "Buffer.drop"
+    else begin
+      b.position <- b.position-len;
```
Comment (dra27) [inline stdlib/buffer.ml]:
```suggestion
      b.position <- b.position - len;
```
(sorry, that's just for consistency with (most of) the rest of the file)

File: stdlib/buffer.mli [hunk 3]
```diff
@@ -158,6 +158,11 @@ val truncate : t -> int -> unit
   Raise [Invalid_argument] if [len < 0] or [len > length b].
   @since 4.05.0 *)
 
+val drop : t -> int -> unit
+(** [drop b len] drops the first [len] bytes of [b]
```
Comment (dra27) [inline stdlib/buffer.mli]:
```suggestion
(** [drop b len] drops the first [len] bytes of [b].
```

**Review by dra27 — COMMENTED** [review]
A few formatting nits, but otherwise LGTM, thanks!

**Comment by alainfrisch** [comment]
Can you provide some intended use cases?  Buffer is mostly an "append-only" structure for now and `truncate` allows to roll back additions, which is useful to "abort a transaction".  I don't see immediately a good use case for `drop`.

In terms of implementation, is reducing the size of the internal `bytes` really necessary/good (it depends on the intended use case), as opposed to just shifting the data within the existing buffer?

**Comment by toots** [comment]
@alainfrisch thanks for asking, I should have added that in the PR's description. We have been using our own (sub-optimal) implementation of this in [savonet/liquidsoap](https://github.com/savonet/liquidsoap) for a while now. See [this link](https://github.com/savonet/liquidsoap/search?q=buffer_drop&unscoped_q=buffer_drop) for detailed usage.

The most generic use is when a `buffer` is used to, ahem, buffer data between two services and sometimes one need to advance the buffer. Typical example recently is an API I wrote to use synchronous read functions in a asynchronous context here: https://github.com/savonet/liquidsoap/blob/external-factorization/src/tools/external_input.ml#L26

As for shortening the buffer, I do agree with you. You must have read some of the earlier commits. In the latest, I use `Bytes.unsafe_blit` which does in-place data move.

**Comment by toots** [comment]
@dra27 Thanks for the review! All the changes have been implemented now.

**Comment by alainfrisch** [comment]
> The most generic use is when a buffer is used to, ahem, buffer data between two services and sometimes one need to advance the buffer.

Aren't you concerned by the cost of copying data in this use case?  Wouldn't other data structures be more suited?  (I'm not opposed to the proposal, just trying to see if it doesn't encourage some use cases which are badly served anyway by the existing data structure.)  At least, what about adding some words to the docstring to explain the cost of the function.

**Comment by toots** [comment]
> > The most generic use is when a buffer is used to, ahem, buffer data between two services and sometimes one need to advance the buffer.
> 
> Aren't you concerned by the cost of copying data in this use case? Wouldn't other data structures be more suited? (I'm not opposed to the proposal, just trying to see if it doesn't encourage some use cases which are badly served anyway by the existing data structure.) At least, what about adding some words to the docstring to explain the cost of the function.

I'm always concerned about performances.. 🙂 Not necessarily in our use-case though. I believe that `memmove` probably provides the best performance for the data move implementation. The other alternative would be to add a `offset` and avoid moving data altogether. I'd be more than happy to look at this implementation but it would have a much bigger impact than the current PR has.

**Comment by bluddy** [comment]
I think that it would make sense to make a RingBuffer implementation for this kind of usecase.

**Comment by toots** [comment]
> I think that it would make sense to make a RingBuffer implementation for this kind of usecase.

The whole `Buffer` module could be implemented using a `RingBuffer` structure. It'd be interesting to ask the original authors why it wasn't done this way, although I'd imagine it's a case of simplicity over performances.

**Comment by bluddy** [comment]
I think it just wasn't needed as part of the design. Buffer is a staging area for concatenating bytes efficiently. Ring buffers are for use-cases where you have a consumer and a producer. I agree that it makes sense to combine the two use-cases, as the cost is minimal.

**Comment by pmetzger** [comment]
I think this is a fine thing addition to the library; sometimes you just want to do things the dumb way because it is "good enough" for your circumstances. It would be neat to get a ring buffer implementation but that need not block this particular PR.

File: stdlib/buffer.mli [hunk 4]
```diff
@@ -158,6 +158,12 @@ val truncate : t -> int -> unit
   Raise [Invalid_argument] if [len < 0] or [len > length b].
   @since 4.05.0 *)
 
+val drop : t -> int -> unit
```
Comment (dbuenzli) [inline stdlib/buffer.mli:161]: Given how buffer are used in practice, I find that to be a confusing name, if I read `Buffer.drop` I might expect it to drop bytes at the end. I would rather suggest `Buffer.drop_{left,right}`.

Comment (toots) [reply stdlib/buffer.mli:161]: How about `advance`?

Comment (dbuenzli) [reply stdlib/buffer.mli:161]: Again I don't find `advanc`ing an append right data structure to be very enlighting. I would prefer to have both `Buffer.drop_left` and `Buffer.drop_right`.

Comment (toots) [reply stdlib/buffer.mli:161]: I'm happy to use whatever has the consensus. `drop_right` would have to be an alias to `truncate`, which already exists. I don't really like aliases personally.

Comment (dbuenzli) [reply stdlib/buffer.mli:161]: No `truncate len` sets the length to `len`. `drop_{left,right} n` remove `n` bytes.

Comment (toots) [reply stdlib/buffer.mli:161]:
```
let drop_right b n =
  truncate b (length b - n)
```
Looks pretty much like an alias to me..

Comment (dbuenzli) [reply stdlib/buffer.mli:161]: Depends on your definition of alias... for me aliases can be substituted one for each other. That is not the case here.

File: stdlib/buffer.ml [hunk 5]
```diff
@@ -350,6 +350,15 @@ let truncate b len =
     else
       b.position <- len
 
+let drop b len =
+    if len < 0 || len > b.position then
+      invalid_arg "Buffer.drop"
```
Comment (dbuenzli) [inline stdlib/buffer.ml:355]: Also it is customary for `drop` functions not to error if you try to drop too much. See e.g. Haskell's `drop`, Batteries `String.{l,r}chop`, Containers `String.drop`. Base `String.drop_{prefix,suffix}`.

**Comment by alainfrisch** [comment]
I feel there should be a more global discussion on whether we want to turn `Buffer` in a generic mutable bytearray/string datastructure, allowing arbitrary in-place edition (including chopping or adding prefix, suffix, or arbitrary segments).  We could do so keeping the current internal definition, at the cost of some (rather efficient) copying; or switch to a different implementation, which could avoid some copying.

**Comment by toots** [comment]
> I feel there should be a more global discussion on whether we want to turn `Buffer` in a generic mutable bytearray/string datastructure, allowing arbitrary in-place edition (including chopping or adding prefix, suffix, or arbitrary segments). We could do so keeping the current internal definition, at the cost of some (rather efficient) copying; or switch to a different implementation, which could avoid some copying.

That's what we ended up doing in liquidsoap: https://github.com/savonet/liquidsoap/blob/master/src/tools/strings.mli
https://github.com/savonet/liquidsoap/blob/master/src/tools/strings.ml

Our implementation could be further optimized by implementing a heuristic to decide wether to copy or not for instance based on the ratio between the string view's length and the containing string's length.

This is just anecdotal & based on my own experience but I wonder how many users of the `Buffer` module use it expecting that kind of optimization under the hood..

**Comment by alainfrisch** [comment]
Thanks for the pointer.  If we had decided to extend Buffer with more general mutation features without changing the internal representation (or perhaps just adding an initial offset to allow allocating free space at the beginning), could this have been a decent replacement for your Strings module (which is based on an entirely different representation, with different performance tradeoffs), in your use case?

**Comment by gasche** [comment]
My understanding is that there is no clear consensus in favor of this change, and that @toots instead went the way of writing a separate library for the more specific need that benefits from a different data-structure.

I see two potential approaches:
1. Try to move forward with the current approach, using @dbuenzli's `drop_{left,right}` suggestion (I agree this improves the proposed API and I think having both would be nice, even if one is related to `truncate`), hoping to convince @alainfrisch that this is a good idea.
2. Close the PR.

If the people involved (@toots in particular) are willing to go for (1), please confirm that this is the intent. Otherwise, without more progress, I will go for (2) next time I cross this still-open PR.

**Comment by toots** [comment]
Hi! I'm happy either way. I believe that the functionality would prove useful to OCaml users in general. I can go with which every naming convention we decide and cleanup the original PR. Just let me know.

**Comment by gasche** [comment]
Maybe @alainfrisch would be the right person to decide which way to take this PR?

**Comment by alainfrisch** [comment]
> Maybe @alainfrisch would be the right person to decide which way to take this PR?

I still find it irregular (design-wise) to add just one operation to an otherwise append-only data structure (+ truncation, which amounts to undoing append operations). There is a valid use case, coherent with the "Buffer" terminology (even though our existing Buffer was really a StringBuilder), but that use case would be even better supported by a different underlying representation.

So I think I would  prefer to keep our Buffer as an append-only "StringBuilder", except if there is a larger consensus on turning it into a more general "mutable & extensible bytearray" datastructure (with insertion on both ends, in-place mutation, etc), perhaps with a different internal representation.

But I've no strong opposition.  If other maintainers have a different view on it, I'll not "fight" against it.

**Comment by gasche** [comment]
Closing, then. This would be better served by an opam package that provides a clean implementation of a bytes ring buffer.

**Pull request closed without merging**